### DISCUSSION TITLE
[@types/three] OBB type missing in v0.150.0

### DISCUSSION THREAD

**Discussion by YordanIliev2002** [DefinitelyTyped/DefinitelyTyped#68973] [Issues with a @types package] [answered]
## Problem
After upgrading `@types/three` from `0.149.0` to `0.150.0`, typescript no longer recognizes `OBB` type definition. Any newer version (e.g. `0.162.0`) also doesn't work. The runtime behavior is unaffected, so it must be an issue with the type.
## Code example
```ts
import { OBB } from "three/examples/jsm/math/OBB";
```

## Error
```
TS2307: Cannot find module 'three/examples/jsm/math/OBB' or its corresponding type declarations.
```

[a comment by an automated account is not shown]

**Comment by Methuselah96** [comment]
Hello @YordanIliev2002!

We [added the `exports` field](https://github.com/three-types/three-ts-types/releases/tag/r150) to the `package.json` in `0.150.0`. Because of this, you will need to update your import to be `import { OBB } from "three/addons/math/OBB.js";` which is [how three.js recommends importing it](https://threejs.org/docs/index.html#manual/en/introduction/Installation).

Let me know if that helped of if you're having another issue.

**Comment by YordanIliev2002** [comment]
Hi, thanks.
Changing the import resulted in `TS1479` error, complaining about ES6 modules vs CommonJS. We can't switch our project to ES6 modules. We are having this issue only with this library, which is leads me to believe that there isn't anything wrong in our configuration. Is there any way that `@types/three` isn't compatible with our import style?

**Reply by Methuselah96** [reply]
Makes sense, can you provide more details about your setup?

- What version of TypeScript are you using?
- What TSConfig settings are you using (e.g., `module`, `moduleResolution`, etc.)?
- Are you using a bundler?

I'm trying to reproduce on my end, but haven't been able to yet.

**Reply by YordanIliev2002** [reply]
The app is a React SPA.

Node 20.x, typescript 5.4.
We are using Webpack, but nothing fancy there.
Those look to be the related options in tsconfig:
```json
{
  "compilerOptions": {
    "target": "ESNext",
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "resolveJsonModule": true,
    .....
    "esModuleInterop": true
  },
  "ts-node": {
    "compilerOptions": {
      ...
      "module": "ESNext",
      "moduleResolution": "Node"
    }
  }
}

```

**Reply by Methuselah96** [reply]
Thanks, I'm able to reproduce now.

Is the file that's importing `OBB` actually being run by Node? The `TS1479` error is correct that you cannot import `OBB` because it is ESM and will crash Node if you try to `require` it. If the file is not being run by Node, maybe you need separate TSConfigs for files that are run by Node vs files that are bundled?

**Reply by YordanIliev2002** [reply]
`OBB` gets imported in a React Component - `Whatever.tsx`, which probably means "node doesn't run the file". We have a single package.json and tsconfig.json, so all code falls under their control. Is there something incompatible in our tsconfig?

**Reply by Methuselah96** [reply, marked as answer]
The problem is that your tsconfig settings are set as if all your files are being run by Node.js, so it's warning you about an import that wouldn't work if you ran it with Node.js.

However, it sounds like you're actually using a bundler for some of your code. With a bundler, importing ESM is perfectly safe. Here are [the recommended settings when using bundlers](https://www.typescriptlang.org/docs/handbook/modules/guides/choosing-compiler-options.html#im-using-a-bundler).

Are your Node.js and React SPA code mixed together or are they in separate directories? If they are separate, you likely want to have separate TSConfigs, one for the React SPA code, and one for the Node.js code. You could then use [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) to type-check both of them with one call of tsc. Or you could just have separate scripts, one to run React SPA type-checking and another to run Node.js type-checking.

**Reply by YordanIliev2002** [reply]
That worked. Turned out that we don't even run ts-node scripts - only plain js. So we didn't even need the two `tsconfig` setup. Thank you so much for your time!
